modules/api_reference/tests/src/Kernel/ParseTest.php

```diff
@@ -0,0 +1,219 @@
+<?php
+
+namespace Drupal\Tests\devportal_api_reference\Kernel;
+
+use Drupal\Core\Cache\NullBackend;
+use Drupal\Core\Logger\LoggerChannelInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\devportal_api_reference\Plugin\Reference\OpenApi3;
+use Drupal\devportal_api_reference\Plugin\Reference\Swagger;
+use Drupal\KernelTests\KernelTestBase;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Tests the parsing and validating edge-case swagger/openapi files.
+ */
+final class ParseTest extends KernelTestBase {
+
+  /**
+   * Mock logger that does nothing.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelInterface
+   */
+  private $mockLogger;
+
+  /**
+   * Mock cache that does nothing.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  private $mockCache;
+
+  /**
+   * Creates a Swagger plugin.
+   *
+   * @return \Drupal\devportal_api_reference\Plugin\Reference\Swagger
+   *   Instance.
+   */
+  private function createSwaggerPlugin(): Swagger {
+    return new Swagger([], '', [], $this->mockCache, $this->mockLogger);
+  }
+
+  /**
+   * Creates an OpenApi3 plugin.
+   *
+   * @return \Drupal\devportal_api_reference\Plugin\Reference\OpenApi3
+   *   Instance.
+   */
+  private function createOpenApi3Plugin(): OpenApi3 {
+    return new OpenApi3([], '', [], $this->mockCache, $this->mockLogger);
+  }
+
+  /**
+   * Lists the files in a subdirectory of fixtures/specials.
+   *
+   * @param string $subdir
+   *   Subdirectory name.
+   *
+   * @return array
+   *   Directory contents in a format that is accepted by PhpUnit's data
+   *   provider. The key is the file name, the value is an array with a single
+   *   element that is the full path of the file.
+   */
+  private function listFiles(string $subdir): array {
+    $dir = realpath(__DIR__ . "/../../fixtures/specials/$subdir");
+
+    $files = [];
+
+    if ($dh = @opendir($dir)) {
```

Comment:
How about using iterators for this? https://www.php.net/manual/en/spl.iterators.php